### ISSUE TITLE
The HLLC Riemann solver fails for some mildly discontinuous problems

### ISSUE THREAD

**Issue by XihZhao** [PrincetonUniversity/athena-public-version#61]
Hi Guys,

I found that the HLLC Riemann Solver in Athena++ sometimes could even fail for a mildly discontinuous Riemann problem. After some examination and comparison with Athena C version, I doubt this issue is related to a few suspicious lines in the hllc.cpp file. Specifically, from line 89 to 107, it defines temp variables ql/qr and used them in the calculation of al/ar, while ql/qr are not contained in the C version HLLC Riemann solver.

My problem generator only contains a simple initial configuration:

```
void MeshBlock::ProblemGenerator(ParameterInput *pin) {
  Real gm1 = peos->GetGamma() - 1.0;
  Real x;
  // Initialize density, momentum
  for (int k=ks; k<=ke; k++){
    for (int j=js; j<=je; j++){
      for (int i=is; i<=ie; i++){
        x=pcoord->x1v(i);

        if (x<0.0){
        phydro->u(IDN,k,j,i) = 0.070;
        phydro->u(IM1,k,j,i) = 3.619*phydro->u(IDN,k,j,i);
        phydro->u(IM2,k,j,i) = -0.570*phydro->u(IDN,k,j,i);
        phydro->u(IM3,k,j,i) = 0.0;
        phydro->u(IEN,k,j,i) =
              0.0340/gm1 +
              (0.5)*
              (SQR(phydro->u(IM1,k,j,i)) + SQR(phydro->u(IM2,k,j,i))
               + SQR(phydro->u(IM3,k,j,i)))/phydro->u(IDN,k,j,i);
        }

        if (x>=0.0){
        phydro->u(IDN,k,j,i) = 4.847;
        phydro->u(IM1,k,j,i) = 0.173*phydro->u(IDN,k,j,i);
        phydro->u(IM2,k,j,i) = 6.584*0.001*phydro->u(IDN,k,j,i);
        phydro->u(IM3,k,j,i) = 0.0;
        phydro->u(IEN,k,j,i) =
              0.242/gm1 +
              (0.5)*
              (SQR(phydro->u(IM1,k,j,i)) + SQR(phydro->u(IM2,k,j,i))
               + SQR(phydro->u(IM3,k,j,i)))/phydro->u(IDN,k,j,i);
        }

        if (MAGNETIC_FIELDS_ENABLED){
        pfield->b.x1f(k,j,i) = 0.0;
        pfield->b.x2f(k,j,i) = 0.0;
        pfield->b.x3f(k,j,i) = 0.0;
        }
      }
    }
  }
  return;
}
```

I compiled it with `python configure.py --prob=RStest --flux=hllc` , and this choice of initial data simply comes from the problem I encountered while running my own simulation.

Now I have tried to solve this Riemann problem by several different solvers: HLLC in Athena++, HLLD with zero-amplitude magnetic field (which is supposed to reduce to HLLC), HLLE, ROE and HLLC solver in Athena C version. Only the HLLC in Athena++ would lead to crash, while all the other solvers worked well. Since the old C version works just fine, I doubt there is an issue in the hllc.cpp file. Could anyone explain where the ql/qr in line 98~101 came from? Do they belong to a standard HLLC algorithm?

**Comment by tomidakn** [comment]
@msbc, can you take a look at this issue?

**Comment by msbc** [comment]
I'll look into this tomorrow. Also those lines come from the general EOS changes (that I made): https://ui.adsabs.harvard.edu/abs/2020ApJS..248....7C/abstract

**Comment by X1Hu1Zhao** [comment]
Okay, thanks @msbc for pointing it out. It is very useful to know about this good work. But I suppose when GENERAL_EOS==0 in Athena++, HLLC should reduce to its original version?

**Comment by msbc** [comment]
Ok, so I recently got an email from Hsi-Yu Schive that I had made a typo when transcribing the equations defining ql/qr to code. Fixing these fixes the issue (I'll post a PR soon). Unrelated, Hsi-Yu also found a bug in the Hydrogen EOS asq function.

Jim, Kengo, and I had a discussion about reducing HLLC to its original version. I had tested various pgens/test with both versions and found no discernible difference in results. So, we came to the conclusion that it was better to have just one version of HLLC and used ther version in my paper.
